server/controllers/stock/reports/stock/avg_med_costs_per_patient_report.js

```diff
@@ -0,0 +1,137 @@
+const {
+  _, ReportManager,
+  STOCK_AVG_MED_COSTS_PER_PATIENT_TEMPLATE,
+  db,
+} = require('../common');
+
+const Exchange = require('../../../finance/exchange');
+
+/**
+ * @method stockAvgMedCostsPerPatientReport
+ *
+ * @description
+ * Construct the report for average medication costs per patient
+ *
+ * GET /reports/stock/avg_med_costs_per_patient'
+ */
+async function stockAvgMedCostsPerPatientReport(req, res, next) {
+  let report;
+  const options = req.query;
+  const {
+    dateFrom, dateTo,
+    depotUuid, depotName,
+    serviceUuid, serviceName,
+  } = options;
+
+  const reportOptions = _.extend({}, options, {
+    filename : 'TREE.AVERAGE_MED_COST_REPORT',
+    title : 'REPORT.AVG_MED_COST_PER_PATIENT.TITLE',
+  });
+
+  const data = {};
+
+  const enterpriseId = req.session.enterprise.id;
+  const exchangeRate = await Exchange.getExchangeRate(enterpriseId, options.currencyId, new Date());
+  const rate = exchangeRate.rate || 1;
+
+  try {
+    report = new ReportManager(STOCK_AVG_MED_COSTS_PER_PATIENT_TEMPLATE, req.session, reportOptions);
+  } catch (e) {
+    return next(e);
+  }
+  const depotSql = depotUuid ? `AND sm.depot_uuid = ?` : '';
+  const serviceSql = serviceUuid ? `AND inv.service_uuid = ?` : '';
+
+  const sql = `
+      SELECT
+        smtot.depot_uuid,  dep.text AS depot_name,
+        smtot.service_uuid, srv.name AS service_name,
+        smtot.srvTotal, smtot.srvPatCount,
+        (smtot.srvTotal / smtot.srvPatCount) AS srvAvgCost,
+        smtot.totalMedCosts, smtot.totalNumPatients,
+        (smtot.totalMedCosts / smtot.totalNumPatients) AS avgMedCosts
+      FROM (
+        SELECT
+          sm.entity_uuid AS patient_uuid,
+          sm.depot_uuid,
+          inv.service_uuid,
+          sm.unit_cost,
+          sm.quantity,
+
+          SUM(sm.quantity * sm.unit_cost) OVER (PARTITION BY sm.depot_uuid, inv.service_uuid) AS srvTotal,
+
+          -- Computes: COUNT(DISTINCT sm.entity_uuid) OVER (PARTITION BY sm.depot_uuid, inv.service_uuid)
+          (DENSE_RANK() OVER (PARTITION BY sm.depot_uuid,
+              inv.service_uuid ORDER BY sm.entity_uuid ASC) +
+           DENSE_RANK() OVER (PARTITION BY sm.depot_uuid,
+              inv.service_uuid ORDER BY sm.entity_uuid DESC) - 1) AS srvPatCount,
```

Comment:
Why the `-1`?  Are we implicitly adding one somewhere?

Reply:
That is the way this work-around works!  See https://stackoverflow.com/questions/11202878/partition-function-count-over-possible-using-distinct

Reply:
I understand - if we can have NULL values, then this is required, presumably to remove the NULL from the count (since its grouped).